### PR TITLE
feat: add experimental checkbox styles

### PR DESCRIPTION
**Pull request by alisonjoseph**
@chrsfshr @joshblack @asudoh 

This is a first pass at adding an experimental component under the feature flag.

- We will need the ibm colors to eventually come from the carbon-elements package (Is that right Josh ?) I just copied the file over for now.

- I removed the dark theme references 

Chris -- things you will need to look out for when copying styles over. You will have to translate anywhere it says `get-fixed-size(3)` to px. Just multiply by 8. so `get-fixed-size(3) becomes `rem(24px)` Not all components have this, but some do (checkbox did). 

If any components have dark theme variants I would just comment that out for now until we determine how we want to handle that. I am assuming for first pass we should just worry about light theme. Not totally sure on this since Security is using dark though? thoughts??

### CODE REVIEW

**Comment by asudoh** [comment]
Great to see the experimental theme for the check box working, and thank you for putting the thoughts here @alisonjoseph! Wrt dark theme - If it can be achieved on top of Carbon (like what we can do for Watson theme), I don’t think not having it in core for now will be a problem.

Looking forward to seeing the finished version!

**Comment by alisonjoseph** [comment]
@asudoh the new testing for styles you added is great! helped me debug. 🙌  🎉

**Comment by asudoh** [comment]
@alisonjoseph The credit [goes to @joshblack](https://github.com/IBM/carbon-components/pull/1021)! 😉

File: src/globals/scss/_feature-flags.scss
Comment (asudoh) [inline src/globals/scss/_feature-flags.scss]: Do we want to do `component-experimental` or `experimental-components` to reflect what this flag does?

**Review by asudoh — APPROVED** [review]
LGTM 👍 bringing up one discussion - Thanks @alisonjoseph!

File: src/globals/scss/_feature-flags.scss
Comment (alisonjoseph) [reply src/globals/scss/_feature-flags.scss]: Thats a good idea 😄

Comment (alisonjoseph) [reply src/globals/scss/_feature-flags.scss]: Maybe we should change it to
```
$experimental-flags (
components: true;
)
```
@joshblack thoughts on this? It's not entirely clear what it's doing right now.

Comment (joshblack) [reply src/globals/scss/_feature-flags.scss]: Feature flags in this case would just correspond to toggling whatever incoming feature we want to test, that was why initially it was `$feature-flags` and `components` since it was toggling the component behavior. Definitely down to have it prefixed by experimental if that would help when reading. If it helps, the mixin to use these flags would be called like:

```scss
@include experimental('components') {
  // ...
}
```

This might help with readability for authors, but I can understand the confusion if we surface this to consumers.

**Comment by joshblack** [comment]
@alisonjoseph do we want to sync up internally to see how we want to do colors? I bet we could get an early publish of the colors package this week if it would help out this work 👍

File: src/globals/scss/_feature-flags.scss
Comment (alisonjoseph) [reply src/globals/scss/_feature-flags.scss]: What about `$experimental-feature-flags` 

I think we should have experimental in there somewhere since that's what consumers will see and that's what we will be calling of these components publicly.

**Comment by alisonjoseph** [comment]
@joshblack yea, we can sync up on the color package, would be helpful to get the actual package linked up.

File: src/globals/scss/_feature-flags.scss
Comment (asudoh) [reply src/globals/scss/_feature-flags.scss]: Thanks @alisonjoseph and @joshblack for your thoughts! Important points from both, the feature flags are for roping things off feature-by-feature basis, and we need to be clear on what incoming features they are. Another thing I’ve been thinking about is releasing roped-off feature should be as simple as flipping the flag.

All that said - Though I brought up “experimental” word, I’m now thinking that the word itself is too general, especially considering another initiative of experimental theme coming after what we are currently working on - Which led me to:

```scss
$feature-flags: (
  components-x: false,
}
```

Though it’s lame. Thoughts on any better words? Thanks!

Comment (alisonjoseph) [reply src/globals/scss/_feature-flags.scss]: If we change it to `components-x` then this part of the code would seems strange. 
```
@include exports('checkbox') {
  @include stable('components') {
    @include checkbox;
  }

  @include experimental('components') {
    @include checkbox--experimental;
  }
}
```

Comment (asudoh) [reply src/globals/scss/_feature-flags.scss]: @alisonjoseph Great point!! What about:

```scss
@include exports('checkbox') {
  @include disabled('components-x') {
    @include checkbox;
  }

  @include enabled('components-x') {
    @include checkbox--experimental;
  }
}
```

BTW as I have signed off this PR, this discussion shouldn’t block this PR to go forward.

Comment (alisonjoseph) [reply src/globals/scss/_feature-flags.scss]: We could discuss at the weekly code review tomorrow with the whole dev team also 😅

Comment (joshblack) [reply src/globals/scss/_feature-flags.scss]: I think the desired API of the mixins should be `stable` and `experimental`, and the string passed in should be the name of the feature flag. Am down to change the name, only used feature flags because historically they are used for experimental/unstable work but am always down for more conciseness here to make sure everything is clear 👍

**Comment by joshblack** [comment]
@alisonjoseph is it possible to get a staging link up and add design as reviewers?

**Comment by alisonjoseph** [comment]
@joshblack the push is failing for me when trying to create a staging link 😢 

I can dig in to it a little more later, as we will need design to look at this stuff at some point.

```
   npm ERR! path /tmp/app/node_modules/loose-envify/cli.js
   npm ERR! code ENOENT
   npm ERR! errno -2
   npm ERR! syscall chmod
   npm ERR! enoent ENOENT: no such file or directory, chmod '/tmp/app/node_modules/loose-envify/cli.js'
   npm ERR! enoent This is related to npm not being able to find a file.
   npm ERR! enoent
   npm ERR! A complete log of this run can be found in:
   npm ERR!     /home/vcap/.npm/_logs/2018-08-30T16_18_49_511Z-debug.log
          **ERROR** Unable to build dependencies: exit status 254
   Failed to compile droplet: Failed to run all supply scripts: exit status 14
   Exit status 223
   Staging failed: STG: Exited with status 223
   Stopping instance c0da0b6e-5bd4-4aba-9e27-2785a7b92224
   Destroying container
   Successfully destroyed container
Error staging application: App staging failed in the buildpack compile phase
FAILED
```

**Comment by joshblack** [comment]
I'm getting a different issue for timing out builds 🤔 cc @asudoh any thoughts?

File: src/components/checkbox/_checkbox.scss
Comment (joshblack) [inline src/components/checkbox/_checkbox.scss]: Do we need to replicate this mixin in the project?

File: src/components/checkbox/_checkbox.scss
Comment (joshblack) [inline src/components/checkbox/_checkbox.scss]: Seems like a strange character before the variable, not sure what it is or if it's from the diff view: `ç$ibm-color__gray-70;`

File: src/globals/scss/_colors--experimental.scss
Comment (joshblack) [inline src/globals/scss/_colors--experimental.scss]: Do we need this file for the experimental work?

File: src/globals/scss/_feature-flags.scss
Comment (joshblack) [inline src/globals/scss/_feature-flags.scss]: Just a reminder to switch this back to `false` 👍

File: src/globals/scss/_functions.scss
Comment (joshblack) [inline src/globals/scss/_functions.scss]: Should we add the SassDoc for this?

File: src/globals/scss/_theme.scss
Comment (joshblack) [inline src/globals/scss/_theme.scss:101]: Would it be helpful to just call this in the experimental mixins themselves instead of toggling at the top-level?

File: src/globals/scss/ibm-colors.scss
Comment (joshblack) [inline src/globals/scss/ibm-colors.scss]: Do you want to try and pull from `@carbon/colors` in this PR?

**Review by joshblack — CHANGES_REQUESTED** [review]
Looks awesome! Just had a couple questions/comments 😄

File: src/components/checkbox/_checkbox.scss
Comment (chrsfshr) [inline src/components/checkbox/_checkbox.scss]: Do I need to add this in the `button.scss` or will this be available to all components once this PR is merged?

File: src/globals/scss/ibm-colors.scss
Comment (alisonjoseph) [reply src/globals/scss/ibm-colors.scss]: Are there any missing colors here? Is there really no yellow? https://github.com/IBM/carbon-elements/blob/master/packages/colors/scss/colors.scss

File: src/components/checkbox/_checkbox.scss
Comment (alisonjoseph) [reply src/components/checkbox/_checkbox.scss]: this is going to get added to all components

File: src/globals/scss/_theme.scss
Comment (alisonjoseph) [reply src/globals/scss/_theme.scss:101]: where do you mean exactly?

**Comment by alisonjoseph** [comment]
any idea why my browser crashes when I try and inspect element? @asudoh @joshblack ?

File: src/globals/scss/_colors--experimental.scss
Comment (alisonjoseph) [reply src/globals/scss/_colors--experimental.scss]: I don't think so, no idea what its even doing.

File: src/components/checkbox/_checkbox.scss
Comment (alisonjoseph) [reply src/components/checkbox/_checkbox.scss]: Oh good catch, can we just switch this to our type scale greek letters? Or do we need it mapped to the a,b,c etc?

Comment (alisonjoseph) [reply src/components/checkbox/_checkbox.scss]: Should we be using the type package?

File: src/components/checkbox/_checkbox.scss
Comment (alisonjoseph) [reply src/components/checkbox/_checkbox.scss]: actually removed this for now I think

File: src/components/checkbox/_checkbox.scss
Comment (alisonjoseph) [reply src/components/checkbox/_checkbox.scss]: updated to use our type-scale for now...

**Comment by asudoh** [comment]
@alisonjoseph Great finding! Looks that it's caused by CSS custom properties (see the scroll bar at the right - very small thumb):

image

Looking at what we can do for that...

**Comment by asudoh** [comment]
Given above screenshot shows multiple instances of [`:root { (CSS custom properties) }` code](https://github.com/IBM/carbon-components/blob/fce07e7962f3c39eca9e7ec26b1cf4a6003233d1/src/globals/scss/ibm-colors.scss#L244-L349), seems that we need to put it in `@include exports()` guard...? Thanks!

**Comment by asudoh** [comment]
@alisonjoseph @joshblack DM'ed wrt pushing to staging server - Hope it helps!

**Comment by alisonjoseph** [comment]
Thanks for finding that @asudoh! was able to remove that code, its the temporary color file I moved over and we don't need all of that anyway.

File: src/globals/scss/ibm-colors.scss
Comment (alisonjoseph) [reply src/globals/scss/ibm-colors.scss]: pulled in the colors, and just added the yellow that was missing. Also just put everything in the colors file since parts of the experimental theme are still using old colors.

**Comment by alisonjoseph** [comment]
Made some updates @joshblack. The only thing this is really missing would be the typescale stuff, and I'd almost say for now just leave ours and add that later?? thoughts? 😭 Although line height seems off in the checkbox with ours.. so maybe we need it right away?

**Comment by aledavila** [comment]
i think since the type scale stuff is still aligned with out it would be ok. It would probably even seem like a "smoother" transition since it would like like not everything changed about the checkbox plus it is experimental.

**Comment by alisonjoseph** [comment]
Ok, added the line height manually. the checkboxes look like they should now 🎉 
screen shot 2018-08-31 at 10 17 57 am

File: src/globals/scss/_colors.scss
Comment (asudoh) [inline src/globals/scss/_colors.scss]: Travis says missing references to this `@import`, what do we want to with this…? Thanks!

Comment (alisonjoseph) [reply src/globals/scss/_colors.scss]: I'm not exactly sure how to fix this, @joshblack ? help? 🆘 😸 

This doesn't work, but I feel like it should since the includePaths are in sass-loader?
`@import '~@carbon/colors/scss/colors';`

Comment (joshblack) [reply src/globals/scss/_colors.scss]: `@import '@carbon/colors/scss/colors';` should do the trick, `~` is a webpack-only thing. Let me know if not!

Comment (alisonjoseph) [reply src/globals/scss/_colors.scss]: Not working 👎 
`Error: File to import not found or unreadable: @carbon/colors/scss/colors`

Comment (alisonjoseph) [reply src/globals/scss/_colors.scss]: nvm. got it working 🎉

**Review by joshblack — APPROVED** [review]

**Pull request merged**
